File: 1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.py
from typing import List
import heapq


class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        cost = 0
        seen = set()
        min_heap = [(0, 0)]

        while len(seen) < n:
            dist, i = heapq.heappop(min_heap)

            if i in seen:
                continue

            seen.add(i)
            cost += dist

            x, y = points[i]

            for j in range(len(points)):
                if j not in seen:
                    other_x, other_y = points[j]
                    man_dist = abs(x - other_x) + abs(y - other_y)
                    heapq.heappush(min_heap, (man_dist, j))

        return cost

File: 1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.java
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Set;

class Solution {

  public int minCostConnectPoints(int[][] points) {
    int n = points.length;
    int cost = 0;
    Set<Integer> seen = new HashSet<>();
    Queue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
    minHeap.add(new int[] { 0, 0 });

    while (seen.size() < n) {
      int[] heapItem = minHeap.poll();
      int dist = heapItem[0];
      int i = heapItem[1];

      if (seen.contains(i)) {
        continue;
      }

      seen.add(i);
      cost += dist;

      int x = points[i][0];
      int y = points[i][1];

      for (int j = 0; j < n; j++) {
        if (!seen.contains(j)) {
          int otherX = points[j][0];
          int otherY = points[j][1];
          int manDist = Math.abs(x - otherX) + Math.abs(y - otherY);
          minHeap.add(new int[] { manDist, j });
        }
      }
    }

    return cost;
  }
}
